### PR TITLE
feat: new generate API and stateful sessions

### PR DESCRIPTION
**Pull request by pavelgj**
`ai.generate` changes

```ts
// can specify a default model
const ai = genkit({
  plugins: [googleAI],
  model: gemini15Flash,
});

// and call generate without specifying the model
const response = await ai.generate('hi');

// toHistory was renamed to (get) messages
const history = response.messages

// can stream as well
const { response, stream } = await ai.generateStream(hi);
for await (const chunk of stream) {
  console.log(chunk.text());
}
console.log((await response).text());
// NOTE: no more `stream()` and `response()`... just `stream` and `response`

// and call generate with just messages (renamed from history)
const response = await ai.generate({
  messages: [ { role: 'user', content: ...} ]
});

// can specify `system` prompt
const response = await ai.generate({
  system: 'talk like a pirate',
  prompt: 'hi',
});
```

simple chat sessions

```ts
const ai = genkit({
  plugins: [googleAI],
  model: gemini15Flash,
});

const chat = await ai.chat();
let response = await chat.send('hi, my name is Pavel');
console.log(response.text()) // "hi Pavel, I'm llm"
// continue the conversation
response = await chat.send('what\'s my name');
console.log(response.text()) // "Pavel"

// can stream
const { response, stream } = await chat.sendStream('bye');
for await (const chunk of stream) {
  console.log(chunk.text());
}
console.log((await response).text());
```

statefull chat sessions

```ts
const ai = genkit({
  plugins: [googleAI],
  model: gemini15Flash,
});

const session = ai.createSession({ store: firestoreSessionStore() });
const chat = await session.chat({ system: 'talk like a pirate '});
let response = await chat.send('hi, my name is Pavel');
console.log(response.text()) // "hi Pavel, I'm llm"
// continue the conversation
response = await chat.send('what\'s my name');
console.log(response.text()) // "Pavel"

// can stream
const { response, stream } = await chat.sendStream('bye');
for await (const chunk of stream) {
  console.log(chunk.text());
}
console.log((await response).text());

// can load session from the store
const prevSession = await ai.loadSession(session.id, { store });
const prevChat = await prevSession.chat();
await prevChat.send('bye');
```

### CODE REVIEW

File: js/ai/src/generate.ts [hunk 1]
```diff
@@ -363,23 +367,36 @@ export class GenerateResponseChunk<T = unknown>
 export async function toGenerateRequest(
```
Comment (mbleigh) [inline js/ai/src/generate.ts:367]: Can you support raw string as input to `generate` as well?

```ts
ai.generate("tell me a long story");
```

(if it's already in here sorry, i missed it)

**Review by mbleigh — COMMENTED** [review]
I don't think I fully understand how chat/session/thread all come together, maybe we can debrief a bit about it soon.

File: js/genkit/src/genkit.ts [hunk 2]
```diff
@@ -667,7 +680,72 @@ export class Genkit {
     if (!resolvedOptions.model) {
       resolvedOptions.model = this.options.model;
     }
-    return runWithRegistry(this.registry, () => generateStream(options));
+    return runWithRegistry(this.registry, () =>
+      generateStream(resolvedOptions)
+    );
+  }
+
+  defineEnvironment<S extends z.ZodTypeAny = z.ZodTypeAny>(config: {
```
Comment (apascal07) [inline js/genkit/src/genkit.ts:688]: Docs on all of these please.

File: js/genkit/src/session.ts [hunk 3]
```diff
@@ -0,0 +1,433 @@
...
+  Genkit,
+  'defineFlow' | 'defineStreamingFlow' | 'defineTool' | 'definePrompt'
+>;
+
+type EnvironmentSessionOptions<S extends z.ZodTypeAny> = Omit<
+  SessionOptions<S>,
+  'store'
+>;
+
+export class Environment<S extends z.ZodTypeAny> implements EnvironmentType {
```
Comment (apascal07) [inline js/genkit/src/session.ts:63]: Doc explaining what it is and how it's used for this and session.

**Review by apascal07 — APPROVED** [review]

File: js/ai/src/generate.ts (hunk 1 above)
Comment (pavelgj) [reply js/ai/src/generate.ts:367]: yes, it's already supported:
https://github.com/firebase/genkit/pull/1019/files#diff-b79c0173e57ae8b5085835ee92582688df1f89dc5e223875cbbc4399d7e790f8R42

File: js/ai/src/generate.ts [hunk 4]
```diff
@@ -412,11 +429,13 @@ export interface GenerateOptions<
 > {
   /** A model name (e.g. `vertexai/gemini-1.0-pro`) or reference. */
   model?: ModelArgument<CustomOptions>;
+  /** The system prompt to be included in the generate request. Can be a string for a simple text prompt or one or more parts for multi-modal prompts (subject to model support). */
+  system?: string | Part | Part[];
   /** The prompt for which to generate a response. Can be a string for a simple text prompt or one or more parts for multi-modal prompts. */
   prompt?: string | Part | Part[];
   /** Retrieved documents to be used as context for this generation. */
   context?: DocumentData[];
```
[the remark this answers is not in the document]
Comment (pavelgj) [reply js/ai/src/generate.ts]: done.

**Comment by pavelgj** [comment]
> I don't think I fully understand how chat/session/thread all come together, maybe we can debrief a bit about it soon.

Updated the API as we discussed.

File: js/genkit/src/chat.ts [hunk 5]
```diff
@@ -0,0 +1,255 @@
...
+  private sessionData?: SessionData<S>;
+  private store: SessionStore<S>;
+  private threadName: string;
+
+  constructor(
+    readonly parent: Genkit | Session<S> | Chat<S>,
+    readonly requestBase?: BaseGenerateOptions,
+    options?: {
+      id?: string;
+      stateSchema?: S;
```
[the remark this answers is not in the document]
Comment (pavelgj) [reply js/genkit/src/chat.ts]: removed state schema... it shouldn't have been there in the first place.

File: js/genkit/src/session.ts [hunk 6]
```diff
@@ -0,0 +1,236 @@
...
+    super(msg);
+  }
+}
+
+/** Session store persists session data such as state and chat messages. */
+export interface SessionStore<S extends z.ZodTypeAny> {
+  get(sessionId: string): Promise<SessionData<S> | undefined>;
+
+  save(sessionId: string, data: SessionData<S>): Promise<void>;
+}
```
[the remark this answers is not in the document]
Comment (pavelgj) [reply js/genkit/src/session.ts:210]: Agreed. SessionStore interface will evolve in the follow up as we add `defineSchema`.

File: js/genkit/src/session.ts [hunk 7]
```diff
@@ -0,0 +1,236 @@
...
+      }
+    );
+  }
+
+  toJSON() {
+    return this.sessionData;
+  }
+}
+
+export interface SessionData<S extends z.ZodTypeAny> {
```
[the remark this answers is not in the document]
Comment (pavelgj) [reply js/genkit/src/session.ts]: done

File: js/genkit/tests/chat_test.ts [hunk 8]
```diff
@@ -0,0 +1,260 @@
...
+      })
+    );
+
+    assert.strictEqual(
+      response.text(),
+      'Echo: hi Genkit; config: {"version":"abc","temperature":11}'
+    );
+  });
+
+  describe('loadChat', () => {
```
[the remark this answers is not in the document]
Comment (pavelgj) [reply js/genkit/tests/chat_test.ts]: simplified to just `loadSession` for now.

**Pull request merged**